chalk-engine/src/logic.rs

```diff
@@ -296,21 +297,23 @@ impl<I: Interner, C: Context<I>> Forest<I, C> {
                 // simplified subgoals. You can think of this as
                 // applying built-in "meta program clauses" that
                 // reduce goals into Domain goals.
-                if let Ok(ex_clause) =
-                    Self::simplify_goal(context, &mut infer, subst, environment, goal)
-                {
-                    info!(
-                        ex_clause = ?infer.debug_ex_clause(context.interner(), &ex_clause),
-                        "pushing initial strand"
-                    );
-                    let strand = Strand {
-                        infer,
-                        ex_clause,
-                        selected_subgoal: None,
-                        last_pursued_time: TimeStamp::default(),
-                    };
-                    let canonical_strand = Self::canonicalize_strand(context, strand);
-                    table.enqueue_strand(canonical_strand);
+                match Self::simplify_goal(context, &mut infer, subst, environment, goal) {
+                    FallibleOrFloundered::Ok(ex_clause) => {
+                        info!(
+                            ex_clause = ?infer.debug_ex_clause(context.interner(), &ex_clause),
+                            "pushing initial strand"
+                        );
+                        let strand = Strand {
+                            infer,
+                            ex_clause,
+                            selected_subgoal: None,
+                            last_pursued_time: TimeStamp::default(),
+                        };
+                        let canonical_strand = Self::canonicalize_strand(context, strand);
+                        table.enqueue_strand(canonical_strand);
+                    }
+                    FallibleOrFloundered::NoSolution => {}
+                    FallibleOrFloundered::Floundered => table.mark_floundered(),
```

Comment:
So, this is somewhat of a question of what to do here. If the goal is a `SubtypeGoal` and both types are `InferenceVar`s, do we want to `Flounder` or have `NoSolution`?

Reply:
Flounder, definitely not `NoSolution`. There is a bit of an interesting problem here around cycles -- I have to remember -- rustc does something special to extend the occurs check to detect cycles like `?A <: ?B` and `?B = Vec<?A>`... not sure if we'll have to reproduce that, but we can worry about that once we get to the generalizer.

Reply:
From what I can tell, rustc has a separate unification table for subtypes that is used only for the occurs check.